### PR TITLE
Gcw 1463 populate organisation table from json data

### PR DESCRIPTION
**Pull request by steveyken**
@trojanh I have some time to review so doing this now

### CODE REVIEW

File: lib/tasks/populate_organisation.rake [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    organisation_type = OrganisationType.find_by_name_en(organisation_type_name) || OrganisationType.create(name_en: organisation_type_name ,name_zh_tw: organisation_type_name ,category_en: organisation_type_name  , category_zh_tw: organisation_type_name)
+
+    #set country
+    country_name = "China - Hong Kong (Special Administrative Region)"
+    country_code = "HKG"
+    country_region = "Asia"
+
+    #for job
+    stockit_id = StockitCountryAddJob.perform_now(country_name, country_name, country_code, country_region)
+    country = Country.find_by(stockit_id: stockit_id)|| Country.create(name_en: country_name ,name_zh_tw: country_name)
```
Comment (steveyken) [inline lib/tasks/populate_organisation.rake]: Can you remove StockitCountryAddJob and replace with

    COUNTRY_NAME_EN = "China - Hong Kong (Special Administrative Region)"
    country = Country.find_by_name_en(COUNTRY_NAME_EN)

File: lib/tasks/populate_organisation.rake [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+namespace :populate_organisation do
+  task organisation:  :environment do
+    organisation_fields_mapping = {
+      "name_en" => "name_en",
+      "name_zh_tw" => "name_zh",
+      "website" => "url",
+    }
+    #set organisation_type
+    organisation_type_name = "NGO"
+    organisation_type = OrganisationType.find_by_name_en(organisation_type_name) || OrganisationType.create(name_en: organisation_type_name ,name_zh_tw: organisation_type_name ,category_en: organisation_type_name  , category_zh_tw: organisation_type_name)
```
Comment (steveyken) [inline lib/tasks/populate_organisation.rake]: When creating OrganisationType, you can leave category_en/zh_tw and name_zh blank. There is no validation for them.

File: lib/tasks/populate_organisation.rake [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    country_name = "China - Hong Kong (Special Administrative Region)"
+    country_code = "HKG"
+    country_region = "Asia"
+
+    #for job
+    stockit_id = StockitCountryAddJob.perform_now(country_name, country_name, country_code, country_region)
+    country = Country.find_by(stockit_id: stockit_id)|| Country.create(name_en: country_name ,name_zh_tw: country_name)
+
+    begin
+      file = Nestful.get("https://goodcitystorage.blob.core.windows.net/public/s88-orgs.json").response.body
```
Comment (steveyken) [inline lib/tasks/populate_organisation.rake]: Can you extract this URL out into a constant and put at the top of the task along with COUNTRY_NAME_EN as well?

File: lib/tasks/populate_organisation.rake [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    rescue Exception => e
+      Airbrake.notify(e, error_class: "populate_organisation", error_message: "Organisation File Error")
```
Comment (steveyken) [inline lib/tasks/populate_organisation.rake]: Good!

File: lib/tasks/populate_organisation.rake [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    # file = File.read('app/assets/organisation.json')
+    if (file.present?)
+      JSON.parse(file).each do |data|
+        organisation_fields_mapping =  organisation_fields_mapping.keep_if { |k, v| data.key? v }
+        organisation = Organisation.find_by(registration: data['org_id']) ||
+          Organisation.new(registration: data['org_id'], organisation_type: organisation_type, country: country)
+          organisation_fields_mapping.each do |organisation_column, data_key|
```
Comment (steveyken) [inline lib/tasks/populate_organisation.rake]: This is good. Can you just fix the indentation slightly so it's easier to read? Thanks

File: app/jobs/stockit_country_add_job.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+class StockitCountryAddJob < ActiveJob::Base
```
Comment (steveyken) [inline app/jobs/stockit_country_add_job.rb]: as discussed, we can delete this job

File: app/services/stockit/country_sync.rb [hunk 7]
```diff
@@ -5,14 +5,35 @@ class CountrySync
 
     include Stockit::Base
```
Comment (steveyken) [inline app/services/stockit/country_sync.rb]: Can we revert these changes as well?

**Comment by steveyken** [comment]
Thanks @trojanh . As we discussed, we'll take a slightly different approach so a few comments here to set that up. Could you also look into writing some tests for the rake task once the changes have been made .. it might be easier to extract the code out into a lib/ class so it's easier to test... we can talk more about this when you're done with the other changes. Thanks. Overall, good job!

**Comment by trojanh** [comment]
Thanks @steveyken for reviewing the PR. I'll make the required changes.

File: lib/tasks/populate_organisation.rake (hunk 3 above)
Comment (namrataukirde) [inline lib/tasks/populate_organisation.rake]: I feel adding constant to Constant.rb will be better. Let me know what you think @steveyken

Comment (steveyken) [inline lib/tasks/populate_organisation.rake]: yes that would make sense too but perhaps if we extract all of this out into a class then the constants can be class constants.

Comment (namrataukirde) [inline lib/tasks/populate_organisation.rake]: Yes Right 👍

File: lib/goodcity/organisation_populator.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
+module Goodcity
```
Comment (steveyken) [inline lib/goodcity/organisation_populator.rb]: @trojanh I'd like to retire the lib/classes folder and switch to using the lib/goodcity namespace so I've moved this class here.

File: lib/tasks/organisations.rake [hunk 9]
```diff
@@ -1,5 +1,12 @@
+require "goodcity/organisation_populator"
```
Comment (steveyken) [inline lib/tasks/organisations.rake]: @trojanh There was already an organisations.rake file so it made sense to put the rake task in there.

File: lib/goodcity/organisation_populator.rb [hunk 10]
```diff
@@ -22,7 +22,7 @@ def initialize
     end
 
 
-    def populate_organisation_db
+    def run
```
Comment (steveyken) [inline lib/goodcity/organisation_populator.rb]: @trojanh I think the class name made it clear what the job is so no need to repeat in the name

File: lib/goodcity/organisation_populator.rb (hunk 8 above)
Comment (trojanh) [inline lib/goodcity/organisation_populator.rb]: Yeah sure, @steveyken . Thanks.

File: lib/goodcity/organisation_populator.rb [hunk 11]
```diff
@@ -12,7 +12,6 @@ class OrganisationPopulator
 
     def initialize
       @organisation_type = get_organisation_type
-      @country           = Country.find_by_name_en(COUNTRY_NAME_EN)
```
Comment (steveyken) [inline lib/goodcity/organisation_populator.rb]: @trojanh In general, the initialize function shouldn't be used to do any 'work' e.g. database calls, http lookups, file loads etc. It should be just enough code to return a new class that hasn't done anything yet but is in a ready and consistent state to be called. E.g. imagine if we added another top level method to count the number of organisations that exist, we wouldn't need to find the country or load the URL just to do this so it would be extra unnecessary work when calling initialize.

File: lib/goodcity/organisation_populator.rb [hunk 12]
```diff
@@ -38,8 +37,9 @@ def run
 
 
     private
-    def get_organisation_type
-      OrganisationType.find_by_name_en(ORGANISATION_TYPE_NAME) || OrganisationType.create(name_en: ORGANISATION_TYPE_NAME)
+
+    def organisation_type
+      @organisation_type ||= OrganisationType.find_or_create_by(name_en: ORGANISATION_TYPE_NAME)
```
Comment (steveyken) [inline lib/goodcity/organisation_populator.rb]: @trojanh It looks like you were looking up the organisation_type in the initializer in order to only do it once and not repeat for each loop iteration. A useful trick here is to use the `||=` operator to memoize the lookup. The first time the function is called, it will call find_or_create_by, but the subsequent times, it will just return the previous (non-nil) value and is therefore effectively cached. This has the advantage of extracting another DB call out of the initializer.

File: lib/goodcity/organisation_populator.rb [hunk 13]
```diff
@@ -10,31 +10,29 @@ class OrganisationPopulator
       "website" => "url",
     }
 
-    def initialize
-      begin
-        @file = Nestful.get(URL).response.body
```
Comment (steveyken) [inline lib/goodcity/organisation_populator.rb]: @trojanh now that we've extracted the URL download out of the initializer there is nothing left to do so it can be removed. Since no initializer is required, we can create a simple `self.run` convenience method so that it's shorter function call i.e. `Goodcity::OrganisationPopulator.run` instead of `Goodcity::OrganisationPopulator.new.run`

Also, the airbrake exceptions should be caught inside the rake task already. If not, then we should add it at the global rake level (for every task) not just this one.

File: spec/lib/goodcity/organisation_populator_spec.rb [hunk 14]
```diff
@@ -4,10 +4,12 @@
 describe Goodcity::OrganisationPopulator do
   let(:organisation_populator) { Goodcity::OrganisationPopulator.new }
   let(:file)                   { File.read("#{Rails.root}/spec/fixtures/organisation.json")}
+  let!(:country)               { FactoryGirl.create(:country, name_en: "China - Hong Kong (Special Administrative Region)") }
```
Comment (steveyken) [inline spec/lib/goodcity/organisation_populator_spec.rb]: @trojanh in general, it's better to use factories to create data and also `let` syntax. (although let! vs before is debatable!)

File: lib/goodcity/organisation_populator.rb (hunk 11 above)
Comment (trojanh) [inline lib/goodcity/organisation_populator.rb]: will keep that in mind henceforth, nice explanation @steveyken

File: spec/lib/goodcity/organisation_populator_spec.rb [hunk 15]
```diff
@@ -29,22 +29,23 @@
   end
 
   context "update organisation" do
+    let(:registration_id_one) { "91/09657"}
```
Comment (steveyken) [inline spec/lib/goodcity/organisation_populator_spec.rb]: @trojanh Use `let` syntax instead of `@instance` variables. `let` has delayed evaluation so it's not run until it's called.

Comment (trojanh) [inline spec/lib/goodcity/organisation_populator_spec.rb]: ok, sure

File: spec/lib/goodcity/organisation_populator_spec.rb [hunk 16]
```diff
@@ -44,12 +44,13 @@
       end
 
       it ":update the existing records" do
+        organisation_populator.run
```
Comment (steveyken) [inline spec/lib/goodcity/organisation_populator_spec.rb]: @trojanh before this change, we were effectively testing that nothing changes when nothing is called! Now we're testing that updates are applied when the populator is run.

**Comment by steveyken** [comment]
@trojanh I've commented on all the commits I've changed so you can get a feel for the sorts of things that would be useful to pick up. Have a read and follow up with me if there are any questions. Otherwise, this is now ready to merge.

**Comment by trojanh** [comment]
I read all your comments and got many things to learn. Thanks for reviewing @steveyken .

File: lib/goodcity/organisation_populator.rb (hunk 10 above)
Comment (trojanh) [inline lib/goodcity/organisation_populator.rb]: I was wondering what name to use, so ended up with that. Anyway run sounds better.
